File: resurch/src/repos/crossref.rs
//! CrossRef API client

use super::traits::{Repository, RepositoryType, SearchResult};
use crate::models::Paper;
use crate::util::{clean_text, extract_doi_from_url, RateLimiter};
use anyhow::{Context, Result};
use async_trait::async_trait;
use reqwest::Client;
use serde::Deserialize;

const CROSSREF_API_BASE: &str = "https://api.crossref.org";

/// CrossRef API client
pub struct CrossRefClient {
    client: Client,
    rate_limiter: RateLimiter,
}

impl CrossRefClient {
    pub fn new() -> Self {
        Self {
            client: Client::builder()
                .user_agent(
                    "resurch/0.1.0 (https://github.com/resurch; mailto:resurch@example.com)",
                )
                .build()
                .expect("Failed to build HTTP client"),
            rate_limiter: RateLimiter::standard_api(),
        }
    }
}

impl Default for CrossRefClient {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl Repository for CrossRefClient {
    fn repository_type(&self) -> RepositoryType {
        RepositoryType::CrossRef
    }

    async fn search(&self, query: &str, offset: u32, limit: u32) -> Result<SearchResult> {
        self.rate_limiter.wait().await;

        let url = format!("{}/works", CROSSREF_API_BASE);

        let response = self
            .client
            .get(&url)
            .query(&[
                ("query", query),
                ("rows", &limit.to_string()),
                ("offset", &offset.to_string()),
            ])
            .send()
            .await
            .context("Failed to send CrossRef search request")?;

        if !response.status().is_success() {
            anyhow::bail!("CrossRef API error: {}", response.status());
        }

        let data: CrossRefSearchResponse = response
            .json()
            .await
            .context("Failed to parse CrossRef response")?;

        let total_results = Some(data.message.total_results as u64);
        let papers: Vec<Paper> = data.message.items.into_iter().map(work_to_paper).collect();

        let has_more = (offset as u64 + papers.len() as u64) < data.message.total_results as u64;

        Ok(SearchResult {
            papers,
            total_results,
            has_more,
            next_offset: if has_more { Some(offset + limit) } else { None },
        })
    }

    async fn get_by_doi(&self, doi: &str) -> Result<Option<Paper>> {
        self.rate_limiter.wait().await;

        let url = format!("{}/works/{}", CROSSREF_API_BASE, doi);

        let response = self
            .client
            .get(&url)
            .send()
            .await
            .context("Failed to send CrossRef DOI request")?;

        if response.status().as_u16() == 404 {
            return Ok(None);
        }

        if !response.status().is_success() {
            anyhow::bail!("CrossRef API error: {}", response.status());
        }

        let data: CrossRefWorkResponse = response
            .json()
            .await
            .context("Failed to parse CrossRef work response")?;

        Ok(Some(work_to_paper(data.message)))
    }

    async fn search_by_title(&self, title: &str) -> Result<Option<Paper>> {
        self.rate_limiter.wait().await;

        let url = format!("{}/works", CROSSREF_API_BASE);

        let response = self
            .client
            .get(&url)
            .query(&[("query.title", title), ("rows", "1")])
            .send()
            .await
            .context("Failed to send CrossRef title search request")?;

        if !response.status().is_success() {
            anyhow::bail!("CrossRef API error: {}", response.status());
        }

        let data: CrossRefSearchResponse = response
            .json()
            .await
            .context("Failed to parse CrossRef response")?;

        Ok(data.message.items.into_iter().next().map(work_to_paper))
    }
}

fn work_to_paper(work: CrossRefWork) -> Paper {
    let title = work.title.into_iter().next().unwrap_or_default();

    let doi = work.doi.clone();
    let doi_url = doi.as_ref().map(|d| format!("https://doi.org/{}", d));

    // Try to extract DOI from URL if not present
    let doi = doi.or_else(|| work.url.as_ref().and_then(|u| extract_doi_from_url(u)));

    let abstract_text = work.r#abstract.map(|a| clean_text(&a));

    let year = work
        .published_print
        .or(work.published_online)
        .or(work.created)
        .and_then(|d| d.date_parts.into_iter().next())
        .and_then(|parts| parts.into_iter().next())
        .map(|y| y as u16);

    let authors = work
        .author
        .unwrap_or_default()
        .into_iter()
        .map(|a| match (a.given, a.family) {
            (Some(given), Some(family)) => format!("{} {}", given, family),
            (None, Some(family)) => family,
            (Some(given), None) => given,
            (None, None) => String::new(),
        })
        .filter(|s| !s.is_empty())
        .collect();

    let publication = work.container_title.into_iter().next();

    Paper {
        id: None,
        doi,
        title,
        title_normalized: None,
        citations: work.is_referenced_by_count.unwrap_or(0) as i32,
        year,
        publication,
        abstract_text,
        snippet: None,
        publisher_url: work.url,
        doi_url,
        source_repository: Some("crossref".to_string()),
        raw_json: None,
        authors,
        publisher: work.publisher,
        created_at: None,
        updated_at: None,
    }
}

// CrossRef API response types

#[derive(Debug, Deserialize)]
struct CrossRefSearchResponse {
    message: CrossRefSearchMessage,
}

#[derive(Debug, Deserialize)]
struct CrossRefSearchMessage {
    #[serde(default)]
    items: Vec<CrossRefWork>,
    #[serde(rename = "total-results", default)]
    total_results: i64,
}

#[derive(Debug, Deserialize)]
struct CrossRefWorkResponse {
    message: CrossRefWork,
}

#[derive(Debug, Deserialize)]
struct CrossRefWork {
    #[serde(rename = "DOI")]
    doi: Option<String>,
    #[serde(default)]
    title: Vec<String>,
    #[serde(rename = "abstract")]
    r#abstract: Option<String>,
    #[serde(rename = "container-title", default)]
    container_title: Vec<String>,
    publisher: Option<String>,
    #[serde(rename = "is-referenced-by-count")]
    is_referenced_by_count: Option<i64>,
    #[serde(rename = "URL")]
    url: Option<String>,
    author: Option<Vec<CrossRefAuthor>>,
    #[serde(rename = "published-print")]
    published_print: Option<CrossRefDate>,
    #[serde(rename = "published-online")]
    published_online: Option<CrossRefDate>,
    created: Option<CrossRefDate>,
}

#[derive(Debug, Deserialize)]
struct CrossRefAuthor {
    given: Option<String>,
    family: Option<String>,
}

#[derive(Debug, Deserialize)]
struct CrossRefDate {
    #[serde(rename = "date-parts", default)]
    date_parts: Vec<Vec<i32>>,
}

File: resurch/resurch/repositories/crossref.py
"""CrossRef repository implementation."""

from typing import AsyncIterator, Optional, List
import httpx

from .base import BaseRepository, PaperResult, SearchProgress
from ..config import API_ENDPOINTS, RESULTS_PER_PAGE, RATE_LIMITS
from ..utils.rate_limiter import get_global_limiter
from ..utils.text import clean_text
from ..utils.doi import normalize_doi


class CrossRefRepository(BaseRepository):
    """CrossRef API repository."""

    name = "crossref"
    default_rate_limit = RATE_LIMITS.get("crossref", 1.0)

    def __init__(self):
        self.base_url = API_ENDPOINTS["crossref"]
        self.per_page = RESULTS_PER_PAGE.get("crossref", 20)

    async def search(
        self,
        query: str,
        max_results: int = 100,
        start_page: int = 0,
        cursor: Optional[str] = None,
    ) -> AsyncIterator[tuple[PaperResult, SearchProgress]]:
        """Search CrossRef for papers matching the query."""
        limiter = get_global_limiter()
        fetched = 0
        page = start_page
        total_results = None

        async with httpx.AsyncClient(timeout=30.0) as client:
            while fetched < max_results:
                await limiter.wait(self.name)

                offset = page * self.per_page
                params = {
                    "query": query,
                    "rows": min(self.per_page, max_results - fetched),
                    "offset": offset,
                }

                try:
                    response = await client.get(self.base_url, params=params)
                    response.raise_for_status()
                    data = response.json()
                except (httpx.HTTPError, ValueError) as e:
                    # Log error but don't crash
                    break

                message = data.get("message", {})
                if total_results is None:
                    total_results = message.get("total-results", 0)

                items = message.get("items", [])
                if not items:
                    break

                for item in items:
                    paper = self._parse_item(item)
                    if paper:
                        fetched += 1
                        progress = SearchProgress(
                            current=fetched,
                            total=min(total_results, max_results) if total_results else None,
                            page=page,
                            is_complete=fetched >= max_results or fetched >= (total_results or max_results),
                        )
                        yield paper, progress

                        if fetched >= max_results:
                            break

                page += 1

                # Check if we've fetched all available results
                if len(items) < self.per_page:
                    break

    async def get_paper_details(self, identifier: str) -> Optional[PaperResult]:
        """Fetch detailed metadata for a paper by DOI."""
        doi = normalize_doi(identifier)
        if not doi:
            return None

        limiter = get_global_limiter()
        await limiter.wait(self.name)

        url = f"{self.base_url}/{doi}"

        async with httpx.AsyncClient(timeout=30.0) as client:
            try:
                response = await client.get(url)
                if response.status_code != 200:
                    return None
                data = response.json()
                return self._parse_item(data.get("message", {}))
            except (httpx.HTTPError, ValueError):
                return None

    def _parse_item(self, item: dict) -> Optional[PaperResult]:
        """Parse a CrossRef item into a PaperResult."""
        if not item:
            return None

        # Get title
        titles = item.get("title", [])
        title = titles[0] if titles else ""
        if not title:
            return None

        title = clean_text(title)

        # Get DOI
        doi = item.get("DOI")

        # Get abstract
        abstract = clean_text(item.get("abstract", ""))

        # Get year
        year = None
        published = item.get("published-print") or item.get("published-online") or item.get("created")
        if published:
            date_parts = published.get("date-parts", [[]])
            if date_parts and date_parts[0]:
                year = date_parts[0][0]

        # Get publication (container title)
        container = item.get("container-title", [])
        publication = container[0] if container else None

        # Get publisher
        publisher = item.get("publisher")

        # Get citation count
        citations = item.get("is-referenced-by-count", 0)

        # Get URLs
        doi_url = f"https://doi.org/{doi}" if doi else None
        publisher_url = item.get("URL")

        # Get PDF URL from links
        pdf_url = None
        links = item.get("link", [])
        for link in links:
            if link.get("content-type") == "application/pdf":
                pdf_url = link.get("URL")
                break

        # Get authors
        authors = []
        for author in item.get("author", []):
            given = author.get("given", "")
            family = author.get("family", "")
            if given and family:
                authors.append(f"{given} {family}")
            elif family:
                authors.append(family)

        return PaperResult(
            title=title,
            doi=doi,
            abstract=abstract,
            year=year,
            publication=publication,
            publisher=publisher,
            citations=citations,
            doi_url=doi_url,
            publisher_url=publisher_url,
            pdf_url=pdf_url,
            authors=authors,
            external_id=doi,
            raw_data=item,
        )
